functionalDB/chapter.md

```diff
@@ -0,0 +1,1097 @@
+
+* Viewed from the outside as a single (i.e., atomic) operation, meaning either all operations succeed or all operations fails
+* Keep the database in a valid state at any given time 
+* Keep the update operations isolated from the outside at any point in time until they are completed. 
+
+The way to answer these requirements is by using a mechanism that takes as an input input the database and a set of operations to be performed, and produces as an output a database whose state reflects the given changes. More than that, all the changes should be seen as an addition of a single layer. 
+
+Performing a transactional operation is done in Clojure using the *Atom* element (this is why we wrap a database with such an element). This element allow us to invoke a single function on it, and all changes to it happen in an atomic and isolated manner. 
+
+As for the function that is to be executed, this function needs to perform all the operations that the user requested, while eventually adding one layer to the database.  To do so we would need to overcome the fact that each of the lifecycle functions adds a layer reflecting the change it inflicts. 
+
+Here we resort to the insight that changes on different attributes (that can be part of either the same or different entities) accumulate when stacking layers, thus a top layer holds all the changes that built the layers below it. Therefore, the solution is to execute each of the user’s operations one after another, each creating a new layer. While doing so, we do not update the databases’s timestamp field. When the last layer is completed, we take only that top layer and place it on the initial database  (leaving all the intermediate layers to pine for the fjords) and only then update the database’s timestamp. All this is happening in the *transact-on-db* function
+
+````clojure
+(defn transact-on-db [initial-db txs]
+     (loop [[tx & rst-tx] txs transacted initial-db]
+       (if tx
+           (recur rst-tx (apply (first tx) transacted (rest tx)))
+           (let [initial-layer  (:layers initial-db)
+                 new-layer (last (:layers transacted))]
+             (assoc initial-db 
+                         :layers (conj  initial-layer new-layer) 
+                         :curr-time (next-ts initial-db) 
+                         :top-id (:top-id transacted))))))
+```` 
+This function is performed on the Atom that wraps the database (to modify the database state as seen by the user), but it can be used also to answer a *what-if* question - as it does not modify the given database, but rather produces a new database with the updated state. These two scenarios are supported by two execution paths that end up in the *transact-on-db* function:

```

Comment:
transact-on-db operates on the database - this is a bug in the writing... I will re-write this section, you are right - it missed the point, especially the fact the both transacting and what-if'ing are two very thin wrappers around the same functionality (which is what happens in transact-on-db), and are doing code reuse via higher order functions and macros.